Cache results in MongoDB

wmanley opened the pull request:
To allow querying the results faster to allow looking up results by job for instance.

**TODO**:
- [x] Write `test_that_generation_id_is_generated_if_not_specified`
- [x] Write `test_that_generation_id_is_persisted_between_invocations`
- [x] Add tests for parsing results directories
- [x] Add tests around mongodb robustness WRT unclean shutdown and restart
- [x] Parse results that are already there in the filesystem but are not in mongo (to cope with unclean shutdown and results generated prior to these changes).
- [x] Factor other UI changes into another pull-request
- [x] Write tests for `AtomicTree.watch`
- [x] Tests: Don't duplicate `find_unused_port_racy`
- [x] Tests: Don't duplicate `MongoDB` test helper
- [x] Fix stb-tester stb-tester/stb-tester#260 to report `test_case`, not `test` in `StateWatch`

**And then in future PRs but not this one**:
- Import results in background, preventing tests from running during this time.
- Merge stb-tester/stb-tester#260
- Make `watch` return a `Watch` object which we can use to unregister a watch and get the initial value.  It should be a `contextmanager`:

        q = Queue()
        def callback(old, new):
            q.push(new)
        with tree.watch(("state", "line_no"), callback) as initial_value:
            send_to_web_ui(inital_value)
            while True:            
                send_to_web_ui(q.pop(block=True))

- Rewrite results view in angular
- Write results into job specific directories
- Add colours to the job progress bar (#20)
- Add endpoint for retrieving results for a given job (particularly the current one)
- Turn off `stbt batch`'s html generating (it's slow in large directories)

drothlis commented on .travis.yml:
```diff
@@ -14,6 +14,7 @@ install:
        isort==3.9.0
        astroid==1.2.1
        pylint==1.3.1
+       pymongo==2.6.3
```
I take it the python-pymongo version (2.1) in Ubuntu 12.04 is too old.

drothlis commented on _stbt_service/results.py:
```diff
@@ -0,0 +1,189 @@
+                    'file': test_file,
+                    'function': test_function
+                },
+                'args': read_file(rundir, 'test-args').split(),
```
`test-args` is newline-separated (to support arguments with spaces).

Annoyingly, `"".split("\n")` returns `[u""]` instead of `[]` so we'll have to handle that.

Also, `read_file` strips all trailing whitespace not just newlines, so an argument of `x   ` will be reported as just `x`.

I'm not sure whether any of this is worth fixing in this pull request.

drothlis commented on _stbt_service/results.py:
```diff
@@ -0,0 +1,196 @@
+    "index.html",
+    "stbt-version.log",
+    "test-args",
+    "test-name"]
```
This blacklist is missing `extra-columns` (compared to `stbt-batch.d/report.py`).

Was this intentional? I suppose that the http service doesn't currently provide a way of setting any extra columns, though you could still write to `extra-columns` in a post-testrun script.

We don't store the extra columns in the database either.

I suppose we can deal with this in a future pull request.

drothlis commented:
>     out = {
>         'start_timestamp': start_timestamp,
>         'end_timestamp': start_timestamp + datetime.timedelta(
>             seconds=float(read_file(rundir, 'duration'))),
>         'test_run': {
>             'test_case': {
>                 'name': test_name,
>                 'file': test_file,
>                 'function': test_function
>             },
>             'args': read_file(rundir, 'test-args').split(),
>             'exit_status': exit_status,
>         },
>         'test_pack': {
>             'revision': read_file(rundir, 'git-commit')
>         },
>         'job': {
>             'category': category
>         },
>         'files': stat_files(rundir)
>     }

Just some misc. observations:

* `stbt run` / `state_watch.py` calls it `test_run.test` instead of `test_run.test_case`. We should probably change `stbt run` so that it is `test_case` in both places.

* `stbt run` also sets `test_run.test.line` (the line number where the Python function starts). I suppose we don't have that information in the results files. If we wanted to show the full testcase script in the results view then we'll have to get `stbt run` to tell `stbt batch run`, which has to then write it to a file... which all seems a lot of hassle. Maybe once we have proved the Mongo solution we can rewrite `stbt batch run` in python and have it insert the results into Mongo directly (with files strictly for backwards compatibility).

wmanley commented:
I've just rebased and squashed.  Rebase dependent branches with:

    git rebase --onto 11760309e8ef22799d047deb45f67b02fc9f9130 \
        95a5a32cddca2bfaadb8bee3e4414ed71f613481
